cpp/include/cudf/io/datasource.hpp

```diff
@@ -335,26 +335,35 @@ class datasource {
   template <typename Container>
   class owning_buffer : public buffer {
    public:
+    // Disallow the container argument to be an lvalue (in which case Container = T& is a
+    // reference).
+    static_assert(!std::is_reference_v<Container>,
+                  "The container argument passed to the constructor must be an rvalue.");
+
     /**
      * @brief Moves the input container into the newly created object.
      *
-     * @param data_owner The container to construct the buffer from (ownership is transferred)
+     * @param moved_data_owner The container to construct the buffer from. Callers should explicitly
+     * pass std::move(data_owner) to this function to transfer the ownership.
      */
-    owning_buffer(Container&& data_owner)
-      : _data(std::move(data_owner)), _data_ptr(_data.data()), _size(_data.size())
+    owning_buffer(Container&& moved_data_owner)
+      : _data(std::forward<Container>(moved_data_owner)),
+        _data_ptr(_data.data()),
+        _size(_data.size())
     {
     }
 
     /**
      * @brief Moves the input container into the newly created object, and exposes a subspan of the
      * buffer.
      *
-     * @param data_owner The container to construct the buffer from (ownership is transferred)
+     * @param moved_data_owner The container to construct the buffer from. Callers should explicitly
+     * pass std::move(data_owner) to this function to transfer the ownership.
      * @param data_ptr Pointer to the start of the subspan
      * @param size The size of the subspan
      */
-    owning_buffer(Container&& data_owner, uint8_t const* data_ptr, size_t size)
-      : _data(std::move(data_owner)), _data_ptr(data_ptr), _size(size)
+    owning_buffer(Container&& moved_data_owner, uint8_t const* data_ptr, size_t size)
+      : _data(std::forward<Container>(moved_data_owner)), _data_ptr(data_ptr), _size(size)
```

Comment:
I'm not sure `std::forward` makes sense here, given that `Container` is never an lvalue reference. How does this fit into the guideline?

Reply:
Perhaps it is simply a matter of preference: The type constraint is only perceived at `static_assert`. For all other parts of the template, I'm still treating `Container&&` as a forwarding reference instead of an rvalue reference.

Reply:
Reverted to using `std::move` for clarity :)